riscv: Optimize bitops and hweight with Zbb extension

uestc-gr opened the pull request:
补丁1：通过 alternative 机制 动态优化位操作函数（如 ffs/fls），在支持 Zbb 扩展的 CPU 上使用指令加速，否则回退到通用 C 实现，并通过编译时常量优化和 NO_ALTERNATIVE 宏确保 EFI 阶段的兼容性。

补丁2：通过Zbb扩展指令集中的cpop/cpopw指令加速hweight()接口的计算

uestc-gr committed:
riscv: Optimize bitops with Zbb extension

mainline inclusion
from mainline-Linux 6.7-rc1
commit 457926b253200bd9bdfae9a016a3b1d1dc661d55
category: feature
bugzilla: https://github.com/RVCK-Project/rvck-olk/issues/96

--------------------------------

This patch leverages the alternative mechanism to dynamically optimize
bitops (including __ffs, __fls, ffs, fls) with Zbb instructions. When
Zbb ext is not supported by the runtime CPU, legacy implementation is
used. If Zbb is supported, then the optimized variants will be selected
via alternative patching.

The legacy bitops support is taken from the generic C implementation as
fallback.

If the parameter is a build-time constant, we leverage compiler builtin to
calculate the result directly, this approach is inspired by x86 bitops
implementation.

EFI stub runs before the kernel, so alternative mechanism should not be
used there, this patch introduces a macro NO_ALTERNATIVE for this purpose.

Signed-off-by: Xiao Wang <xiao.w.wang@intel.com>
Reviewed-by: Charlie Jenkins <charlie@rivosinc.com>
Link: https://lore.kernel.org/r/20231031064553.2319688-3-xiao.w.wang@intel.com
Signed-off-by: Palmer Dabbelt <palmer@rivosinc.com>
Signed-off-by: Gao Rui <gao.rui@zte.com.cn>

oervci commented:
开始测试

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/78/consoleFull

oervci commented:
Kernel build success!

uestc-gr committed:
riscv: Optimize hweight API with Zbb extension

mainline inclusion
from mainline-Linux 6.8-rc1
commit 55ca8d7aa2af3ebdb6f85cccf1b0703d031c1678
category: feature
bugzilla: https://github.com/RVCK-Project/rvck-olk/issues/96

--------------------------------

The Hamming Weight of a number is the total number of bits set in it, so
the cpop/cpopw instruction from Zbb extension can be used to accelerate
hweight() API.

Signed-off-by: Xiao Wang <xiao.w.wang@intel.com>
Reviewed-by: Charlie Jenkins <charlie@rivosinc.com>
Link: https://lore.kernel.org/r/20231112095244.4015351-1-xiao.w.wang@intel.com
Signed-off-by: Palmer Dabbelt <palmer@rivosinc.com>
Signed-off-by: Gao Rui <gao.rui@zte.com.cn>

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/83/consoleFull

oervci commented:
开始测试

oervci commented:
check patch done. log: https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/check-patch/79/consoleFull

oervci commented:
Kernel build success!

oervci commented:
kunit test done. log:https://jenkins.oerv.ac.cn/job/rvck-pipeline/job/kunit-test/85/consoleFull

oervci commented:
Lava check done! result url: https://lava.oerv.ac.cn/scheduler/job/242
